### ISSUE TITLE
Remove usages of `Course.program` and `Course.position_in_program`

### ISSUE THREAD

**Issue by rhysyngsun** [mitodl/mitxonline#1220]
Going forward we'll be using `ProgramRequirement` to determine which courses are in a program and what their order is. This means we should be removing these old fields so that we won't have issues with the same data being in two possible places (or at worst, these two pieces of data disagreeing).

**Comment by rachellougee** [comment]
@rhysyngsun Just to confirm this means we are dropping these two fields from the program table? I will work on getting these two fields replaced with 'ProgramRequirement' as these fields are staged in our data platform https://github.com/mitodl/ol-data-platform/issues/381

**Comment by rhysyngsun** [comment]
@rachellougee yeah, eventually

**Comment by pdpinch** [comment]
@rhysyngsun have you started on this? If not, I'd like to remove your assignment and put it into the backlog. I don't think it's particularly urgent.

**Comment by asadali145** [comment]
There are a number of properties like `first_unexpired_run` and `first_course_unexpired_runs` that are using the `Course.program` relationship. We have not replaced that with `ProgramRequirement`. What do we need to do with such properties/usages?

**Comment by asadali145** [comment]
There are a couple of usages of `position_in_program` like sorting courses and in a property `first_unexpired_run`. What do we need to do in these cases?

**Comment by rhysyngsun** [comment]
@asadali145 It should use the `ProgramRequirement` table going forward, I suspect in many cases we may not actually want to be sorting by `position_in_program` unless it's to display the entire set of courses to a learner. See below for details.

I glanced over the code:

- `Program.next_run_date` - this doesn't appear to be used and to be honest the implementation doesn't seem to return a particularly useful value, I'd say let's remove this.
- `Program.first_course_unexpired_runs` - same as above
- `Program.first_unexpired_run` - as you mentioned, used in the CMS, as with the above fields, the implementation takes the first run of the first course ordered by `position_in_program`, which isn't necessarily the actual first unexpired run across all courses in that program, so that logic is wrong to begin with. The order in the program shouldn't be used to determine the closest course run chronologically. I think what you'd be looking for is something roughly like this:
```python
CourseRun.objects.filter(course__in_programs__program=program).order_by("start_date").first()
```

**Comment by asadali145** [comment]
Hi @pdpinch @rhysyngsun @jkachel,

There are some cases where we are using `Course.program` like in `cms.models.FlexiblePricingRequestForm.get_parent_courseware`. With new program requirements, a course can be associated with multiple programs. `get_parent_courseware` Is used to determine financial assistance tiers and is also used to check FA submissions. How do we need to handle this with FA and other places (if any)?

**Comment by asadali145** [comment]
Similarily, `Course.program` is also used to generate Program Certificates. Using the CourseRunEnrollment, we get the associated program of a course and return the program certificate. Which program should be returned in case a course is part of multiple programs?

**Comment by pdpinch** [comment]
> With new program requirements, a course can be associated with multiple programs. get_parent_courseware Is used to determine financial assistance tiers and is also used to check FA submissions. How do we need to handle this with FA and other places (if any)?

This is a really good question about Financial Assistance, which is actually a product question. I'll have to get back to you on the intent.

**Comment by asadali145** [comment]
> I don't understand the context of when this would be called. Can you give some more details?

@pdpinch User dashboard displays the program enrollments of a user. These program enrollments are implicit with course run enrollments i.e. if a user registers in a course/course run then we display the associated program in the dashboard. We do this by doing `CourseRunEnrollment.course.program`.

In case a course is added to multiple programs, Are we going to display all the associated programs in the dashboard?

**Comment by asadali145** [comment]
> This is a really good question about Financial Assistance, which is actually a product question. I'll have to get back to you on the intent.

@pdpinch Do you have an update on this one?

**Comment by pdpinch** [comment]
> This is a really good question about Financial Assistance, which is actually a product question. I'll have to get back to you on the intent.
> 
> Do you have an update on this one?

Not yet. What's easiest?

**Comment by asadali145** [comment]
> Not yet. What's simplest?

@pdpinch I am not sure about the simplest because it seems like the current implementation is very tightly coupled with the `Course.program`.

One idea that we discussed (Me & @arslanashraf7 ) is that we could leave this to the learner.

**APPROACH**: Add an option for the learner to choose the program for Financial Assistance(FA). This can be done by adding a new field in the Financial Assistance form.

**_CON:_** We associate the FA forms with the programs in the CMS. So, we generate FA form links like `{program_readable_id}/{FA_FORM_SLUG}`.

**_AN ISSUE:_** How do we decide which FA form link we need to display to the user?
**_A POSSIBLE SOLUTION:_** Associate the FA forms with the courses only and generate links like `{course_readable_id}/{FA_FORM_SLUG}`. Then we can add a new choice field in the FA Form where users can choose the desired program. **_This would require us to manage forms for the individual courses._** 


@rhysyngsun @jkachel What do you guys think about this? Do you have any better idea to handle FA when a course is required for multiple programs?

**Comment by jkachel** [comment]
I think there's an option similar to adding a field to the form - we could pop a modal or put a page in between to have them choose the program first, then send them to the appropriate financial assistance form. 

I specifically don't like moving the forms to the courses; that seems like it'd result in a mess of similar forms (all the DEDP courses now share the same form, for instance) and I'm not sure it solves the problem, since there'd still be the question of who should process the request for the learner if the course is in >1 program.

**Comment by rhysyngsun** [comment]
Do we currently use the financial aid for courses form (I suspect we don't)? If not, could we just disable this feature for now (and file an issue to fix it) until we have a need for it and know what we want to do with this?

**Comment by jkachel** [comment]
> Do we currently use the financial aid for courses form (I suspect we don't)?

[We do now, actually.](https://github.com/mitodl/hq/issues/774)

**Comment by rhysyngsun** [comment]
So I think given that, we should do this:

- If the course isn't in a program, there's nothing to do/show
- If the course is only in 1 program, use that, but don't show any UI to the learner
- If the course is in multiple programs, display a select option for the learner to pick which one they want to apply to

**Comment by pdpinch** [comment]
Sorry, there are still some unanswered questions about courses in multiple programs, and how financial aid should be handled. 

Can we put this on hold until I can get those answered? Or should we move ahead with this and open a follow-up issue fo the UI?

There's a good chance that we will have our first courses in multiple programs this fall.

**Comment by asadali145** [comment]
> Can we put this on hold until I can get those answered? Or should we move ahead with this and open a follow-up issue fo the UI?

@pdpinch I think we can put this on hold until we have clear answers. It is better to make all the changes in a single go.

Also, The PR https://github.com/mitodl/mitxonline/pull/1429 for the removal of `Course.position_in_program` is ready and is almost in the approval stage. This is independent of the `Course.program` as well. Should we go ahead and merge this PR once approved or do we need to wait?

**Comment by pdpinch** [comment]
Thanks all for your patience. I've discussed this with various stakeholders, and we've decided that the personal pricing acceptance should apply to all the programs the course is a member of. 

Here's a concrete example: 

1. Learner goes to about page for 14.100x. It's a member of program 1 (Develpment Economics) and program 2 (US Economics)
2. They click the link to request financial aid and fill out the form. Their country & income combination requires approval. 
3. A team member of program 1 or program 2 approve it. 
4. Now the learner gets the personal price for any course that is a member of program 1 or program 2. 

> Also, The PR https://github.com/mitodl/mitxonline/pull/1429 for the removal of Course.position_in_program is ready and is almost in the approval stage. This is independent of the Course.program as well. Should we go ahead and merge this PR once approved or do we need to wait?

Based on your comment, I think this personal pricing discussion needs a new issue, so we can merge #1429 and close this. Let me know if you disagree.
